core/src/banking_stage/thread_aware_account_locks.rs

```diff
@@ -0,0 +1,663 @@
+        self.schedulable_threads_with_read_only_handler(account, |thread_set, counts| {
+            thread_set
+                .only_one_scheduled()
+                .filter(|thread_id| counts[*thread_id] < self.sequential_queue_limit)
+                .map_or_else(ThreadSet::none, ThreadSet::only)
+        })
+    }
+
+    /// Returns `ThreadSet` of schedulable threads, given the read-only lock handler.
+    /// Helper function, since the only difference between read and write schedulable threads
+    /// is in how the case where only read locks are held is handled.
+    /// If there are no locks, then all threads are schedulable.
+    /// If only write-locked, then only the thread holding the write lock is schedulable.
+    /// If a mix of locks, then only the write thread is schedulable.
+    /// The sequential limit is checked, and a thread will not be returned as schedulable
+    /// if the limit is reached.
+    fn schedulable_threads_with_read_only_handler(
+        &self,
+        account: &Pubkey,
+        read_only_handler: impl Fn(&ThreadSet, &[u32]) -> ThreadSet,
+    ) -> ThreadSet {
+        match (self.write_locks.get(account), self.read_locks.get(account)) {
+            (None, None) => ThreadSet::any(self.num_threads),
+            (None, Some((thread_set, counts))) => read_only_handler(thread_set, counts),
+            (Some((thread_id, count)), None) => {
+                if count == &self.sequential_queue_limit {
+                    ThreadSet::none()
+                } else {
+                    ThreadSet::only(*thread_id)
+                }
+            }
+            (Some((thread_id, count)), Some((thread_set, counts))) => {
+                debug_assert_eq!(Some(*thread_id), thread_set.only_one_scheduled());
+                if count + counts[*thread_id] == self.sequential_queue_limit {
+                    ThreadSet::none()
+                } else {
+                    ThreadSet::only(*thread_id)
+                }
+            }
+        }
+    }
+
+    /// Add locks for all writable and readable accounts on `thread_id`.
+    pub fn lock_accounts<'a>(
+        &mut self,
+        write_account_locks: impl Iterator<Item = &'a Pubkey>,
+        read_account_locks: impl Iterator<Item = &'a Pubkey>,
+        thread_id: ThreadId,
+    ) {
+        assert!(thread_id < self.num_threads);
+        for account in write_account_locks {
+            self.write_lock_account(account, thread_id);
+        }
+
+        for account in read_account_locks {
+            self.read_lock_account(account, thread_id);
+        }
+    }
+
+    /// Locks the given `account` for writing on `thread_id`.
+    /// Panics if the account is already locked for writing on another thread.
+    fn write_lock_account(&mut self, account: &Pubkey, thread_id: ThreadId) {
+        match self.write_locks.entry(*account) {
+            std::collections::hash_map::Entry::Occupied(mut entry) => {
```

Comment:
nit: maybe add `use ...` at top of this file so that this code becomes just `Occupied` or `Entry::Occupied`.

Reply:
263b7e3b7d9367b4a5ca44d73499766ae5485f68